pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+	"k8s.io/klog"
+)
+
+type DualWriterMode2 struct {
+	DualWriter
+}
+
+// NewDualWriterMode2 returns a new DualWriter in mode 2.
+func NewDualWriterMode2(legacy LegacyStorage, storage Storage) *DualWriterMode2 {
+	return &DualWriterMode2{*newDualWriter(legacy, storage)}
+}
+
+// Get overrides the default behavior of the Storage and retrieves an object from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriterMode2) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Getter)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Getter is missing")
+	}
+
+	s, err := d.Storage.Get(ctx, name, &metav1.GetOptions{})
+	if err == nil {
+		return s, err
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	return legacy.Get(ctx, name, &metav1.GetOptions{})
+}
+
+// List overrides the default behavior of the Storage and retrieves objects from
+// LegacyStorage or Storage depending on the DualWriter mode.
+func (d *DualWriterMode2) List(ctx context.Context, options *metainternalversion.ListOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Lister)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Lister is missing")
+	}
+
+	legacyList, err := legacy.List(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+	ll, err := meta.ExtractList(legacyList)
+	if err != nil {
+		return nil, err
+	}
+
+	storageList, err := d.Storage.List(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+	sl, err := meta.ExtractList(storageList)
+	if err != nil {
+		return nil, err
+	}
+
+	type fieldsToEnrich struct {
+		resourceVersion string
+	}
+
+	m := map[string]fieldsToEnrich{}
+	for _, obj := range sl {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		m[accessor.GetName()] = fieldsToEnrich{accessor.GetResourceVersion()}
+	}
+
+	for _, obj := range ll {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return nil, err
+		}
+		if entry, ok := m[accessor.GetName()]; ok {
+			accessor.SetResourceVersion(entry.resourceVersion)
+		}
+	}
+
+	if err = meta.SetList(legacyList, ll); err != nil {
+		return nil, err
+	}
+	return legacyList, nil
+}
+
+// Create overrides the default behavior of the Storage and writes to LegacyStorage and Storage depending on the dual writer mode.
+func (d *DualWriterMode2) Create(ctx context.Context, obj runtime.Object, createValidation rest.ValidateObjectFunc, options *metav1.CreateOptions) (runtime.Object, error) {
+	legacy, ok := d.legacy.(rest.Creater)
+	if !ok {
+		return nil, fmt.Errorf("legacy storage rest.Creater is missing")
+	}
+
+	created, err := legacy.Create(ctx, obj, createValidation, options)
+	if err != nil {
+		return created, err
+	}
+
+	accessor, err := meta.Accessor(created)
+	if err != nil {
+		return created, err
+	}
+	accessor.SetResourceVersion("")
```

Comment:
Why are we setting a new object to an empty resource_version?

Reply:
I'm a bit fuzzy on this now because it's been a while and I tried playing with this in other methods as well. I think it has to do with the Create method expecting the object to be new and therefore not having a resource version set. Because legacy Create converts the playllist entity to a resource, it does set a resource version if I remember correctly.

Reply:
Found the spot. It's a requirement as implied by [this error](https://github.com/grafana/grafana/blob/main/pkg/apiserver/storage/file/file.go#L38).

Reply:
oh, I see, it seems it happens because of this part [here](https://github.com/grafana/grafana/blob/main/pkg/apiserver/storage/file/file.go#L161-L167). Thanks!